lambdas/indexer/app.py

```diff
@@ -484,3 +484,43 @@ def nudge(event: chalice.app.CloudWatchEvent):
         log.info('Nudging queue with %i token(s) for a total value of %i', len(tokens), num_documents)
         for batch in chunked(tokens, 10):
             _send_tokens(batch)
+
+
+@app.schedule('rate(5 minutes)', name=config.failure_message_lambda_basename)
+def retrieve_failure_messages(_event: chalice.app.CloudWatchEvent):
+    """
+    Get all the messages from the fail queue and move these messages to the the DynamoDB failure message table.
+    """
+    database = boto3.resource('dynamodb')
+    table = database.Table(config.dynamo_failure_message_table_name)
+    with table.batch_writer() as writer:
+        fail_queue = queue(config.fail_queue_name)
+        messages = []
+        while True:
+            received_messages = fail_queue.receive_messages(AttributeNames=['All'],
+                                                            MaxNumberOfMessages=10,
+                                                            WaitTimeSeconds=20,
+                                                            VisibilityTimeout=300)
+            if received_messages:
+                messages.extend(received_messages)
+            if len(messages) >= 10 or (len(messages) > 0 and not received_messages):
+                batch_requests = []
+                for message in messages[:10]:
+                    body = json.loads(message.body)
+                    writer.put_item(Item={
+                        'MessageType': 'bundle_notification' if 'notification' in body.keys() else 'other',
+                        'SentTimeMessageId': f"{message.attributes['SentTimestamp']}-{message.message_id}",
+                        'Body': message.body
+                    })
+                    batch_requests.append({
+                        'Id': message.message_id,
+                        'ReceiptHandle': message.receipt_handle,
+                        'VisibilityTimeout': 0
+                    })
+                fail_queue.change_message_visibility_batch(Entries=batch_requests)
+                for message in batch_requests:
+                    del message['VisibilityTimeout']
+                fail_queue.delete_messages(Entries=batch_requests)
+                del messages[:10]
```

Comment:
I think this is overly complicated. You can deal with each batch separately there is no need to put message in a global list and then remove at the end. If you get only three messages back from SQS its ok to only write three  messages to DynanmoDb.